File: frontend/src/api/contractor.ts
import { http, unwrap } from './http'
import type { PageResult, WorkoverMeasure } from '../types/workover'

export type ContractorStatus = 'AVAILABLE' | 'BUSY' | 'OFFLINE' | 'EXCEPTION'
export type ContractorSourceType = 'EXTERNAL_SYNC' | 'LOCAL_SUPPLEMENT' | 'SYNC_ERROR'
export type ContractorSyncStatus = 'SYNCED' | 'PENDING_CONFIRM' | 'CONFLICT' | 'INVALID'
export type ContractorSyncResultStatus = 'SUCCESS' | 'FAILED' | 'PARTIAL'
export type ContractorSyncType = 'SCHEDULED' | 'MANUAL' | 'SINGLE_TEAM'
export type OperationStatus = 'WAITING_DISPATCH' | 'DISPATCHED' | 'WORKING' | 'FINISHED' | 'CANCELED'

export interface ContractorCapacity {
  id: number
  contractor_name: string
  team_name: string
  report_date: string
  available_count: number
  status: ContractorStatus
  capability_tags: Record<string, unknown>
  external_system_id?: string | null
  external_status?: string | null
  source_type: ContractorSourceType
  sync_status: ContractorSyncStatus
  last_synced_at?: string | null
  sync_error_message?: string | null
  confirmed_at?: string | null
  confirmed_by_id?: number | null
  contact_name?: string | null
  contact_phone?: string | null
  qualification_expire_at?: string | null
  equipment_summary?: string | null
  occupied_count: number
  created_at: string
  updated_at: string
}

export interface ContractorCapacityPayload {
  contractor_name: string
  team_name: string
  report_date: string
  available_count: number
  status: ContractorStatus
  capability_tags: Record<string, unknown>
  external_system_id?: string | null
  external_status?: string | null
  source_type?: ContractorSourceType
  sync_status?: ContractorSyncStatus
  sync_error_message?: string | null
  contact_name?: string | null
  contact_phone?: string | null
  qualification_expire_at?: string | null
  equipment_summary?: string | null
}

export interface ContractorSyncLog {
  id: number
  sync_type: ContractorSyncType
  status: ContractorSyncResultStatus
  started_at: string
  finished_at?: string | null
  success_count: number
  failed_count: number
  created_count: number
  updated_count: number
  ignored_count: number
  error_message?: string | null
  operator_id?: number | null
  raw_summary: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface ContractorSyncSummary {
  connection_status: string
  last_sync_time?: string | null
  last_sync_status?: ContractorSyncResultStatus | null
  created_count: number
  updated_count: number
  ignored_count: number
  failed_count: number
  exception_count: number
}

export interface ContractorOverview {
  reported_team_count: number
  available_team_count: number
  busy_team_count: number
  offline_team_count: number
  sync_exception_count: number
  major_repair_team_count: number
}

export interface ContractorOperationLink {
  id: number
  operation_no: string
  status: OperationStatus
  well_no?: string | null
  dispatch_time?: string | null
  a5_status?: string | null
  created_at: string
}

export interface OperationSheet {
  id: number
  project_id: number
  project_well_no?: string | null
  project?: {
    id: number
    well_no?: string | null
    block_name?: string | null
    territory_unit?: string | null
    report_unit?: string | null
    data_source?: string | null
    measures_jsonb?: { measures?: WorkoverMeasure[] }
    approved_at?: string | null
  } | null
  contractor_capacity_id: number | null
  contractor?: {
    id: number
    contractor_name: string
    team_name: string
    report_date?: string | null
    status?: ContractorStatus | string
  } | null
  operation_no: string
  status: OperationStatus
  planned_start_at?: string | null
  planned_end_at?: string | null
  actual_start_at?: string | null
  actual_end_at?: string | null
  progress: number
  progress_detail: Record<string, unknown>
  a5_status?: string | null
  a5_remark?: string | null
  last_a5_sync_at?: string | null
  created_at: string
  updated_at: string
}

export interface DispatchA5Result {
  sheet: OperationSheet
  redirect_url: string
  token: string
  expire_at: string
}

export interface ContractorQuery {
  page?: number
  page_size?: number
  contractor_name?: string
  team_name?: string
  report_date?: string
  status?: ContractorStatus | ''
  capability_tag?: string
  source_type?: ContractorSourceType | ''
  sync_status?: ContractorSyncStatus | ''
}

export interface OperationSheetQuery {
  page?: number
  page_size?: number
  status?: OperationStatus | ''
  project_id?: number
  contractor_capacity_id?: number
  well_no?: string
  block_name?: string
  contractor_keyword?: string
  start_date?: string
  end_date?: string
}

function compact(params: object) {
  return Object.fromEntries(Object.entries(params).filter(([, value]) => value !== '' && value !== undefined && value !== null))
}

export function listContractors(query: ContractorQuery) {
  return unwrap<PageResult<ContractorCapacity>>(http.get('/contractors/', { params: compact(query) }))
}

export function getContractorSyncSummary() {
  return unwrap<ContractorSyncSummary>(http.get('/contractors/sync-summary'))
}

export function getContractorOverview(reportDate?: string) {
  return unwrap<ContractorOverview>(http.get('/contractors/overview', { params: compact({ report_date: reportDate }) }))
}

export function syncContractors(payload: { report_date?: string }) {
  return unwrap<ContractorSyncLog>(http.post('/contractors/sync', payload))
}

export function syncSingleContractor(id: number) {
  return unwrap<ContractorSyncLog>(http.post(`/contractors/${id}/sync`))
}

export function confirmContractor(id: number) {
  return unwrap<ContractorCapacity>(http.patch(`/contractors/${id}/confirm`))
}

export function markContractorException(id: number, reason: string) {
  return unwrap<ContractorCapacity>(http.patch(`/contractors/${id}/exception`, { reason }))
}

export function resolveContractorException(id: number) {
  return unwrap<ContractorCapacity>(http.patch(`/contractors/${id}/resolve-exception`))
}

export function listContractorSyncLogs(query: { page?: number; page_size?: number; status?: ContractorSyncResultStatus | ''; sync_type?: ContractorSyncType | '' }) {
  return unwrap<PageResult<ContractorSyncLog>>(http.get('/contractors/sync-logs', { params: compact(query) }))
}

export function listContractorOperationSheets(id: number) {
  return unwrap<ContractorOperationLink[]>(http.get(`/contractors/${id}/operation-sheets`))
}

export function createContractor(payload: ContractorCapacityPayload) {
  return unwrap<ContractorCapacity>(http.post('/contractors/', payload))
}

export function updateContractor(id: number, payload: Partial<ContractorCapacityPayload>) {
  return unwrap<ContractorCapacity>(http.put(`/contractors/${id}`, payload))
}

export function listOperationSheets(query: OperationSheetQuery) {
  return unwrap<PageResult<OperationSheet>>(http.get('/contractors/operation-sheets/', { params: compact(query) }))
}

export function listPrioritySheets() {
  return unwrap<OperationSheet[]>(http.get('/contractors/priority-sheets'))
}

export function createOperationSheet(payload: { project_id: number; planned_start_at?: string; planned_end_at?: string }) {
  return unwrap<OperationSheet>(http.post('/contractors/operation-sheets/', payload))
}

export function dispatchOperation(payload: { operation_sheet_id: number; contractor_capacity_id: number; redirect_path?: string }) {
  return unwrap<DispatchA5Result>(http.patch('/contractors/dispatch', payload))
}

export function updateOperationProgress(id: number, payload: { progress: number; progress_detail: Record<string, unknown> }) {
  return unwrap<OperationSheet>(http.patch(`/contractors/operation-sheets/${id}/progress`, payload))
}

export interface OperationAnalytics {
  total_sheets: number
  status_distribution: {
    waiting_dispatch: number
    dispatched: number
    working: number
    finished: number
    canceled: number
  }
  dispatch_rate: number
  completion_rate: number
  team_workload: { team_name: string; sheet_count: number }[]
  measure_type_distribution: { measure_type: string; count: number }[]
  anomaly_count: number
}

export function getOperationAnalytics() {
  return unwrap<OperationAnalytics>(http.get('/contractors/analytics/summary'))
}

File: app/schemas/contractor.py
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.workover import (
    ContractorCapacitySourceType,
    ContractorCapacityStatus,
    ContractorCapacitySyncResultStatus,
    ContractorCapacitySyncStatus,
    ContractorCapacitySyncType,
    OperationStatus,
)


class ContractorCapacityCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contractor_name: str = Field(min_length=1, max_length=128)
    team_name: str = Field(min_length=1, max_length=128)
    report_date: date
    available_count: int = Field(default=0, ge=0)
    status: ContractorCapacityStatus = ContractorCapacityStatus.AVAILABLE
    capability_tags: dict[str, Any] = Field(min_length=1)
    sync_error_message: str | None = Field(default=None, max_length=500)
    contact_name: str | None = Field(default=None, max_length=64)
    contact_phone: str | None = Field(default=None, max_length=32)
    qualification_expire_at: date | None = None
    equipment_summary: str | None = None

    @field_validator("capability_tags")
    @classmethod
    def validate_capability_tags(cls, value: dict[str, Any]) -> dict[str, Any]:
        if not any(item is True or item == "true" or item == 1 for item in value.values()):
            raise ValueError("capability_tags must include at least one enabled capability")
        return value


class ContractorCapacityUpdate(BaseModel):
    """人工维护基础资料。

    运力数量、队伍状态、来源类型和同步状态由补录、同步确认、异常处理、派工/释放等专门流程维护。
    """

    contractor_name: str | None = Field(default=None, min_length=1, max_length=128)
    team_name: str | None = Field(default=None, min_length=1, max_length=128)
    report_date: date | None = None
    capability_tags: dict[str, Any] | None = None
    external_status: str | None = Field(default=None, max_length=64)
    sync_error_message: str | None = None
    contact_name: str | None = Field(default=None, max_length=64)
    contact_phone: str | None = Field(default=None, max_length=32)
    qualification_expire_at: date | None = None
    equipment_summary: str | None = None

    @field_validator("capability_tags")
    @classmethod
    def validate_optional_capability_tags(cls, value: dict[str, Any] | None) -> dict[str, Any] | None:
        if value is not None and not any(item is True or item == "true" or item == 1 for item in value.values()):
            raise ValueError("capability_tags must include at least one enabled capability")
        return value


class ContractorCapacityQuery(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=200)
    contractor_name: str | None = None
    team_name: str | None = None
    report_date: date | None = None
    status: ContractorCapacityStatus | None = None
    capability_tag: str | None = None
    source_type: ContractorCapacitySourceType | None = None
    sync_status: ContractorCapacitySyncStatus | None = None


class ContractorCapacityOut(BaseModel):
    id: int
    contractor_name: str
    team_name: str
    report_date: date
    available_count: int
    status: ContractorCapacityStatus
    capability_tags: dict[str, Any]
    external_system_id: str | None = None
    external_status: str | None = None
    source_type: ContractorCapacitySourceType
    sync_status: ContractorCapacitySyncStatus
    last_synced_at: datetime | None = None
    sync_error_message: str | None = None
    confirmed_at: datetime | None = None
    confirmed_by_id: int | None = None
    created_by_id: int | None = None
    contact_name: str | None = None
    contact_phone: str | None = None
    qualification_expire_at: date | None = None
    equipment_summary: str | None = None
    occupied_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ContractorSyncPayload(BaseModel):
    report_date: date | None = None


class ContractorExceptionPayload(BaseModel):
    reason: str = Field(min_length=1, max_length=500)


class ContractorCapacitySyncLogQuery(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=200)
    status: ContractorCapacitySyncResultStatus | None = None
    sync_type: ContractorCapacitySyncType | None = None


class ContractorCapacitySyncLogOut(BaseModel):
    id: int
    sync_type: ContractorCapacitySyncType
    status: ContractorCapacitySyncResultStatus
    started_at: datetime
    finished_at: datetime | None = None
    success_count: int
    failed_count: int
    created_count: int
    updated_count: int
    ignored_count: int
    error_message: str | None = None
    operator_id: int | None = None
    raw_summary: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ContractorSyncSummary(BaseModel):
    connection_status: str
    last_sync_time: datetime | None = None
    last_sync_status: ContractorCapacitySyncResultStatus | None = None
    created_count: int = 0
    updated_count: int = 0
    ignored_count: int = 0
    failed_count: int = 0
    exception_count: int = 0


class ContractorCapacityOverview(BaseModel):
    reported_team_count: int
    available_team_count: int
    busy_team_count: int
    offline_team_count: int
    sync_exception_count: int
    major_repair_team_count: int


class ContractorOperationSheetLinkOut(BaseModel):
    id: int
    operation_no: str
    status: OperationStatus
    well_no: str | None = None
    dispatch_time: datetime | None = None
    a5_status: str | None = None
    created_at: datetime


class WorkoverOperationSheetCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_id: int = Field(ge=1)
    planned_start_at: datetime | None = None
    planned_end_at: datetime | None = None


class WorkoverOperationSheetUpdate(BaseModel):
    contractor_capacity_id: int | None = None
    planned_start_at: datetime | None = None
    planned_end_at: datetime | None = None
    actual_start_at: datetime | None = None
    actual_end_at: datetime | None = None
    progress: int | None = Field(default=None, ge=0, le=100)
    progress_detail: dict[str, Any] | None = None


class WorkoverOperationSheetQuery(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=200)
    status: OperationStatus | None = None
    project_id: int | None = None
    contractor_capacity_id: int | None = None
    well_no: str | None = None
    block_name: str | None = None
    contractor_keyword: str | None = None
    start_date: date | None = None
    end_date: date | None = None


class WorkoverOperationSheetOut(BaseModel):
    id: int
    project_id: int
    project_well_no: str | None = None
    contractor_capacity_id: int | None = None
    operation_no: str
    status: OperationStatus
    planned_start_at: datetime | None = None
    planned_end_at: datetime | None = None
    actual_start_at: datetime | None = None
    actual_end_at: datetime | None = None
    progress: int
    progress_detail: dict[str, Any]
    a5_status: str | None = None
    a5_remark: str | None = None
    last_a5_sync_at: datetime | None = None
    last_a5_report_date: date | None = None
    a5_sync_result: str | None = None
    a5_sync_error: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DispatchPayload(BaseModel):
    operation_sheet_id: int = Field(ge=1)
    contractor_capacity_id: int = Field(ge=1)
    redirect_path: str = Field(default="/measure-review", min_length=1, max_length=128)


class DispatchA5Out(BaseModel):
    sheet: WorkoverOperationSheetOut
    redirect_url: str
    token: str
    expire_at: datetime


class ProgressPatch(BaseModel):
    progress: int = Field(ge=0, le=100)
    progress_detail: dict[str, Any] = Field(default_factory=dict)
